### PR TITLE
Tighten index buffer null behavior

### PR DESCRIPTION
**Pull request by HansKristian-Work**

### CODE REVIEW

File: libs/vkd3d/command.c [hunk 1]
```diff
@@ -5440,7 +5454,8 @@ static void STDMETHODCALLTYPE d3d12_command_list_IASetIndexBuffer(d3d12_command_
 
     if (!view)
     {
-        WARN("Ignoring NULL index buffer view.\n");
+        WARN("Got NULL index buffer view, indexed draw calls will be dropped.\n");
```
Comment (doitsujin) [inline libs/vkd3d/command.c:5457]: Does this actually unbind the current index buffer or should we just ignore the call entirely?

Comment (HansKristian-Work) [reply libs/vkd3d/command.c:5457]: Can test

File: libs/vkd3d/command.c [hunk 2]
```diff
@@ -3678,6 +3679,19 @@ static void STDMETHODCALLTYPE d3d12_command_list_DrawIndexedInstanced(d3d12_comm
             iface, index_count_per_instance, instance_count, start_vertex_location,
             base_vertex_location, start_instance_location);
 
+    if (!list->has_valid_index_buffer)
+    {
+        FIXME_ONCE("Application attempts to perform an indexed draw call without index buffer bound.\n");
+        /* We are supposed to render all 0 indices here. However, there are several problems with emulating this approach.
+         * There is no robustness support for index buffers, and if we render all 0 indices,
+         * it is extremely unlikely that this would create a meaningful side effect.
```
Comment (Joshua-Ashton) [inline libs/vkd3d/command.c:3687]: gl_VertexIndex to draw quads/tris?

Comment (HansKristian-Work) [reply libs/vkd3d/command.c:3687]: gl_VertexIndex is always going to return 0 since the indices are 0.

File: libs/vkd3d/command.c (hunk 1 above)
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:5457]: It does unbind the index buffer, yes.

File: libs/vkd3d/command.c [hunk 3]
```diff
@@ -5455,15 +5470,21 @@ static void STDMETHODCALLTYPE d3d12_command_list_IASetIndexBuffer(d3d12_command_
             index_type = VK_INDEX_TYPE_UINT32;
             break;
         default:
-            WARN("Invalid index format %#x.\n", view->Format);
-            return;
+            FIXME_ONCE("Invalid index format %#x. This will map to R16_UINT to match observed driver behavior.\n", view->Format);
+            /* D3D12 debug layer disallows this case, but it doesn't trigger a DEVICE LOST event, so we shouldn't crash and burn.
+             * Mark the command buffer as not having a valid index buffer. */
```
Comment (doitsujin) [inline libs/vkd3d/command.c:5475]: > `* Mark the command buffer as not having a valid index buffer.

That's not what we're doing here, though.

File: tests/d3d12.c [hunk 4]
```diff
@@ -35726,6 +35726,152 @@ static void test_vertex_shader_stream_output(bool use_dxil)
...
+    pso_desc.StreamOutput.NumEntries = ARRAY_SIZE(so_declaration);
+    pso_desc.StreamOutput.pSODeclaration = so_declaration;
+    pso_desc.StreamOutput.pBufferStrides = strides;
+    pso_desc.StreamOutput.NumStrides = ARRAY_SIZE(strides);
+    pso_desc.StreamOutput.RasterizedStream = D3D12_SO_NO_RASTERIZED_STREAM;
+    hr = ID3D12Device_CreateGraphicsPipelineState(device, &pso_desc,
+            &IID_ID3D12PipelineState, (void **)&context.pipeline_state);
+    if (hr == E_NOTIMPL)
+    {
+        skip("Stream output is not supported.\n");
```
Comment (doitsujin) [inline tests/d3d12.c:35794]: Are we ever going to run into this?

**Review by doitsujin — CHANGES_REQUESTED** [review]
One code comment seems wrong, other than that, looking good.

File: tests/d3d12.c (hunk 4 above)
Comment (HansKristian-Work) [reply tests/d3d12.c:35794]: Copy-paste from the vertex buffer streamout test.

File: libs/vkd3d/command.c (hunk 3 above)
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:5475]: Derp.

Comment (HansKristian-Work) [reply libs/vkd3d/command.c]: Fixed.

File: tests/d3d12.c (hunk 4 above)
Comment (doitsujin) [reply tests/d3d12.c:35794]: Doesn't really matter either way.

**Review by doitsujin — APPROVED** [review]

**Pull request merged**